File: src/lib/worktree.ts
import { execFile as execFileCb } from "node:child_process";
import { promisify } from "node:util";
import { BUILD_TIMEOUT } from "./types";

const execFile = promisify(execFileCb);

/** Create a new worktree via wt-create. */
export async function create(name: string, branch?: string): Promise<string> {
  const args = ["--non-interactive", "--worktree-name", name];
  if (branch) args.push(branch);

  const { stdout } = await execFile("wt-create", args, {
    timeout: BUILD_TIMEOUT,
  });
  return stdout.trim();
}

/** List all worktrees via wt-list. */
export async function list(): Promise<string> {
  const { stdout } = await execFile("wt-list", [], { timeout: BUILD_TIMEOUT });
  return stdout.trim();
}

/** Delete a worktree via wt-delete. */
export async function remove(name: string): Promise<string> {
  const { stdout } = await execFile("wt-delete", [name], {
    timeout: BUILD_TIMEOUT,
  });
  return stdout.trim();
}

/** Open a worktree via wt-open. */
export async function open(name: string): Promise<string> {
  const { stdout } = await execFile("wt-open", [name], {
    timeout: BUILD_TIMEOUT,
  });
  return stdout.trim();
}

File: packages/api/internal/worktree/worktree.go
package worktree

import (
	"context"
	"os/exec"
	"strings"
	"time"
)

const buildTimeout = 30 * time.Second

// Create creates a new worktree via wt-create.
func Create(name string, branch string) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), buildTimeout)
	defer cancel()

	args := []string{"--non-interactive", "--worktree-name", name}
	if branch != "" {
		args = append(args, branch)
	}

	cmd := exec.CommandContext(ctx, "wt-create", args...)
	out, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(out)), nil
}

// List lists all worktrees via wt-list.
func List() (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), buildTimeout)
	defer cancel()

	cmd := exec.CommandContext(ctx, "wt-list")
	out, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(out)), nil
}

// Delete deletes a worktree via wt-delete.
func Delete(name string) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), buildTimeout)
	defer cancel()

	cmd := exec.CommandContext(ctx, "wt-delete", name)
	out, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(out)), nil
}

// Open opens a worktree via wt-open.
func Open(name string) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), buildTimeout)
	defer cancel()

	cmd := exec.CommandContext(ctx, "wt-open", name)
	out, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(out)), nil
}
